### PR TITLE
Bug 1933711: EgressDNS: Keep short lived records at most 30s

### PR DESCRIPTION
**Pull request by juanluisvaladas**
The TTL is guaranteed to be valid for its duration, however, it's not
guaranteed that the DNS server will cache it and always return the same
list of IP addreses for the whole TTL.

For records with large TTL (>30 min) we assume they will always return
the same list of addresses and query them only once every 30 min. We'll
assume rest of the records a TTL of at most 30 seconds.

### CODE REVIEW

**Comment by juanluisvaladas** [comment]
/retitle Bug 1933711: EgressDNS: Keep short lived records at most 30s

File: pkg/network/common/dns.go [hunk 1]
```diff
@@ -400,3 +400,20 @@ func removeDuplicateIPs(ips []net.IP) []net.IP {
 
 	return uniqueIPs
 }
+
+func normalizeTTL(ttl time.Duration) time.Duration {
+
+	// The TTL is guaranteed to be valid for its duration, however, it's not
```
Comment (danwinship) [inline pkg/network/common/dns.go:406]: (extra blank line)

File: pkg/network/common/dns.go [hunk 2]
```diff
@@ -400,3 +400,20 @@ func removeDuplicateIPs(ips []net.IP) []net.IP {
...
+	// guaranteed that the DNS server will cache it and always return the same
+	// list of IP addreses for the whole TTL.
+	// For records with large TTL (>30 min) we assume they will always return
+	// the same list of addresses and query them only once every 30 min. We'll
+	// assume rest of the records a TTL of at most 30 seconds.
+
+	if ttl < 30*time.Second {
+		return ttl
+	} else if ttl >= 1800*time.Second {
+		return 1800 * time.Second
```
Comment (danwinship) [inline pkg/network/common/dns.go:416]: seriously, this is what gofmt expects? No wonder I can never figure out if it wants spaces or not...

anyway, `30 * time.Minute` would be clearer

File: pkg/network/common/dns_test.go [hunk 3]
```diff
@@ -160,8 +160,9 @@ func TestAddDNS(t *testing.T) {
 			if !ipsEqual(d.ips, test.ips) {
 				t.Fatalf("Test case: %s failed, expected IPs: %v, got: %v for the domain %q", test.testCase, test.ips, d.ips, test.domainName)
 			}
-			if d.ttl.Seconds() != test.ttl {
-				t.Fatalf("Test case: %s failed, expected TTL: %g, got: %g for the domain %q", test.testCase, test.ttl, d.ttl.Seconds(), test.domainName)
+			normalizedTTL := normalizeTTL(time.Duration(test.ttl) * time.Second)
```
Comment (danwinship) [inline pkg/network/common/dns_test.go:163]: If you made `test.ttl` a `time.Duration` you could get rid of a lot of casting/multiplying

**Review by danwinship — COMMENTED** [review]
looks good. just nitpicks

File: pkg/network/common/dns.go (hunk 2 above)
Comment (juanluisvaladas) [reply pkg/network/common/dns.go]: Yes, this is generated by gofmt :sweat_smile:

File: pkg/network/common/dns_test.go [hunk 4]
```diff
@@ -508,6 +513,29 @@ func TestUpdateDNS(t *testing.T) {
...
+	tests[31] = 30
+	tests[1799] = 30
+	tests[1800] = 1800
+	tests[1801] = 1800
+	tests[3600] = 1800
+	for k, v := range tests {
+		originalTTL := time.Duration(k) * time.Second
+		expectedTTL := time.Duration(v) * time.Second
+		normalizedTTL := normalizeTTL(originalTTL)
+		if normalizedTTL != time.Duration(v)*time.Second {
```
Comment (abhat) [inline pkg/network/common/dns_test.go:533]: Use `expectedTTL` for comparison.

**Comment by abhat** [comment]
Mostly looks good. One minor comment.

**Comment by danwinship** [comment]
/lgtm
i guess the `time.Duration` change didn't really save much, it just moved the messiness around

**Pull request merged**